### DISCUSSION TITLE
RFC: Built-in markdown / mdx content editor

### DISCUSSION THREAD

**Discussion by timlrx** [timlrx/pliny#34] [Ideas]
## Built-in markdown / mdx content editor 

Building on next.js makes for a very interesting possibility of coupling the project with a built-in markdown / mdx content editor. It will take the form of an API route which integrates with the markdown schema as specified in `contentlayer.config.ts` and a given set of MDX plugins. This opens 2 modes of using it:

1. Local editing - modified content is saved to your local drive
2. Hosted CMS - a user can access the API route, edit content and automatically commit it to git like [forestry.io](https://forestry.io/)

Thanks to contentlayer, there's no need to re-invent the wheel and require the user to re-specify the frontmatter schema like other CMSes out there. We can also use it for validation. 

On the editor side, using [Milkdown](https://github.com/Saul-Mirone/milkdown) which is built on prosemirror and remark/rehype makes integration much more straightforward since everything can be transformed using a similar remark/rehype pipeline specified in `contentlayer.config`. It also has numerous plugins which gives it a Medium like experience. Check out the [online demo](https://milkdown.dev/online-demo).

## Implementation

There are 3 main screens on the interface side:

1. Overall content view, displaying all the files in content directory
2. Editor view which allows editing frontmatter information and md/mdx body
3. Preview mode which renders the content as it would look like when published 

It will be published as a standalone package within the monorepo (@pliny/editor) and can be integrated with other nextjs, contentlayer projects out there. 

Milkdown already has existing integration with certain remark libraries like prismjs for code editing and katex for math. Beyond that, integration with specific remark/rehype packages and adding MDX components might require more work to define and scope out how it would happen and how the user could actually interact with it.

**Comment by AustenLamacraft** [comment]
Is the idea that each instance will be effectively be running its own CMS via Next's API routes? This is then different from [TinaCMS](https://tina.io/) (successor to forestry.io) who are offering their own Git-backed CMS.

I think this may hit a sweet spot for people who are happy to have a GH account to store their content but don't want to get their hands dirty with committing it.

There's also [motif](https://motif.land/) which allows you to work with locally stored content, which is pretty cool.
